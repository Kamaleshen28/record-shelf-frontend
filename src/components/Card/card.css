
.card{
    height: 180px;
    border-radius: 10px;
    box-shadow: 3px 3px rgba(0.1, 0.1, 0.1, 0.1);
    padding: 10px;
    display: flex;
    flex-direction: column;
    gap:20px;
}
.light-card{
    background-color: #DFDFDF;
}

.dark-card{
    background-color: #ABABAB;
}

@media only screen and (min-width: 530px) {  
    .card { 
        width: 180px;
    }
  }
@media only screen and (max-width: 529px) {  
    .card { 
        width:240px;
    }
  }

.card-image{
    width: 110px;
    border-radius: 50%;
}

.details-container{
    display: flex;
    align-items: center;
    justify-content: space-between;
}


.text-container{
    display: flex;
    flex-direction: column;
    align-items: flex-start;
}

.song-name-text{
    font-size: 20px;
    font-weight: lighter;
    color: black;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 1;
    overflow: hidden;
    justify-content: flex-start;
    text-align: left;

}

.song-genre-text{
    font-weight: bold;
    font-size: 14px;
}
.heart-image-container{
    position: relative;
}
.count-text{
    position: absolute;
    right: 0;
    left:0;
    top:7px;
    cursor: pointer;
}
.heart-image{
    cursor: pointer;
}