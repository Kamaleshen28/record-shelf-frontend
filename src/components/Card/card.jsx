import React, { useEffect, useState } from 'react'
import './card.css'
import heartRed from '../../assets/heart-red.svg'
import heartGray from '../../assets/heart-gray.svg'
import axios from 'axios'

export default function Card (props) {

    const [like, setLike] = useState();
    const [count, setCount] = useState();

    useEffect(() => {
        fetchData()
    }, [])
    const fetchData =  async() => {
        const songData = await axios.get(`http://localhost:8080/api/records/${props.id}/likes`, {
            headers: {
              'Authorization': 'Basic QWlzaHdhcnlhIE4='
            }
        })
        setLike(songData.data.data.like)
        setCount(songData.data.data.count)
    }

    const updateLikeCount = async (data) => {
        const config = {
            method: 'patch',
            url: `http://localhost:8080/api/records/${props.id}/likes`,
            headers: { 
              'Authorization': 'Bearer QWlzaHdhcnlhIE4=', 
              'Content-Type': 'application/json'
            },
            data : data
        };
        const response = await axios(config)
        setCount(response.data.data.count)
    }

    const onHeartClickHandler = () => {
        setLike(previousValue => {
            if(previousValue){
                const data = JSON.stringify({"like": !previousValue})
                updateLikeCount(data)
            }else{
                const data = JSON.stringify({"like": !previousValue})
                updateLikeCount(data)
            }
            return !previousValue;
        })
    }
    
    return(
        <div className={(props.lightCard)? 'card light-card': 'card dark-card'}>
            <div className='card-image-container'>
                <img className='card-image' src={props.imageUrl} alt='Not Found' />
            </div>
            <div className='details-container'>
                <div className='text-container'>
                    <a className='song-name-text'>{props.name}</a>
                    <a className='song-genre-text'>{props.genre.name}</a>
                </div>
                <div className='heart-image-container'>
                    <img className='heart-image' onClick={onHeartClickHandler} src={like? heartRed: heartGray} alt='Not Found' />
                    <a className='count-text' onClick={onHeartClickHandler} >{count}</a>
                </div>
            </div>
        </div>
    )
}