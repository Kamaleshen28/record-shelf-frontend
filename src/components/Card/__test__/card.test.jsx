import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import Card from '../card';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';



const mockSongData = 
{
    'id': '0e2ddccf-f494-4c41-a499-90f8267f491a',
    'name': 'Beautiful Mistakes (feat. Megan Thee Stallion)',
    'genre': {
        'id': '128aa7f8-c943-48ce-b352-7edd26fa4c6e',
        'name': 'Pop'
    },
    'artist': {
        'id': '496b0a85-2bfa-45bc-8d0f-57fe0ce55708',
        'name': 'Maroon 5'
    },
    'imageUrl': 'https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631',
    'publishedAt': '2021-06-11T00:00:00'
};

const mockLightCard = false;

const mockLikeDataBeforeClick = {data :{
    'like': false,
    'count': 0
}
};
const mockLikeDataAfterClick = {data :{
    'like': true,
    'count': 1
}
};

const mockLikeDataBeforeClickTrue = {data :{
    'like': true,
    'count': 1
}
};
const mockLikeDataAfterClickFalse = {data :{
    'like': false,
    'count': 0
}
};

describe('card', () => {
    it('should render card correctly', async () =>{
        const {asFragment} = render(<Card {...mockSongData} lighCard={mockLightCard} />);
        await waitFor(() =>{
            expect(screen.getByText('Pop')).toBeTruthy();
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it('should toggle the heart image on when heart image is clicked', async () =>{
        render(<Card {...mockSongData} lighCard={mockLightCard} />);
        await waitFor(() =>{
            expect(screen.getByText('Pop')).toBeTruthy();
        });
        const getHeartImage = screen.queryByTestId('heart-image');
        // const heartSrcAttribute = getHeartImage.getAttribute('src');
        fireEvent.click(getHeartImage);
        expect(getHeartImage.getAttribute('src')).toEqual('heart-red.svg');
    });

    it('should toggle the heart image on when heart image is clicked', async () =>{
        jest.spyOn(axios, 'get').mockResolvedValue({data:mockLikeDataBeforeClick});
        render(<Card {...mockSongData} lighCard={mockLightCard} />);
        await waitFor(() =>{
            expect(screen.getByText('0')).toBeTruthy();
        });
        const getHeartImage = screen.queryByTestId('heart-image');
        const likeCounter = screen.queryByTestId('like-count');
        const mock = new MockAdapter(axios);
        mock.onPatch('http://localhost:8080/api/records/0e2ddccf-f494-4c41-a499-90f8267f491a/likes').reply(200, mockLikeDataAfterClick);
        fireEvent.click(getHeartImage);
        await waitFor(() =>{
            expect(screen.getByText('1')).toBeTruthy();
        });
        expect(likeCounter.textContent).toEqual('1');
    });
    it('should toggle the heart image on when heart image is clicked', async () =>{
        jest.spyOn(axios, 'get').mockResolvedValue({data:mockLikeDataBeforeClick});
        render(<Card {...mockSongData} lighCard={mockLightCard} />);
        await waitFor(() =>{
            expect(screen.getByText('0')).toBeTruthy();
        });
        const likeCounter = screen.queryByTestId('like-count');
        const mock = new MockAdapter(axios);
        mock.onPatch('http://localhost:8080/api/records/0e2ddccf-f494-4c41-a499-90f8267f491a/likes').reply(200, mockLikeDataAfterClick);
        fireEvent.click(likeCounter);
        await waitFor(() =>{
            expect(screen.getByText('1')).toBeTruthy();
        });
        expect(likeCounter.textContent).toEqual('1');
    });

    it('should toggle the heart image on when heart image is clicked', async () =>{
        jest.spyOn(axios, 'get').mockResolvedValue({data:mockLikeDataBeforeClickTrue});
        render(<Card {...mockSongData} lighCard={mockLightCard} />);
        await waitFor(() =>{
            expect(screen.getByText('1')).toBeTruthy();
        });
        const likeCounter = screen.queryByTestId('like-count');
        const mock = new MockAdapter(axios);
        mock.onPatch('http://localhost:8080/api/records/0e2ddccf-f494-4c41-a499-90f8267f491a/likes').reply(200, mockLikeDataAfterClickFalse);
        fireEvent.click(likeCounter);
        await waitFor(() =>{
            expect(screen.getByText('0')).toBeTruthy();
        });
        expect(likeCounter.textContent).toEqual('0');
    });
});