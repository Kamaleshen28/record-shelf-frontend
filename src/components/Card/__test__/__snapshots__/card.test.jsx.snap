// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card should render card correctly 1`] = `
<DocumentFragment>
  <div
    class="card dark-card"
  >
    <div
      class="card-image-container"
    >
      <img
        alt="Not Found"
        class="card-image"
        src="https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631"
      />
    </div>
    <div
      class="details-container"
    >
      <div
        class="text-container"
      >
        <a
          class="song-name-text"
        >
          Beautiful Mistakes (feat. Megan Thee Stallion)
        </a>
        <a
          class="song-genre-text"
        >
          Pop
        </a>
      </div>
      <div
        class="heart-image-container"
      >
        <img
          alt="Not Found"
          class="heart-image"
          data-testid="heart-image"
          src="heart-gray.svg"
        />
        <a
          class="count-text"
          data-testid="like-count"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
